BUG: Include broadcasting for ``rtol`` argument in ``matrix_rank``

mtsokol opened the pull request:
Hi @asmeurer,

This PR fixes broadcasting of `rtol` argument in `matrix_rank` function that you pointed out in https://github.com/numpy/numpy/pull/25437#discussion_r1500043325.

mtsokol committed:
FIX: Include broadcasting for rtol argument

rgommers commented:
Thanks! A regression test would be great indeed.

mhvk commented on numpy/linalg/_linalg.py:
```diff
@@ -2068,18 +2069,23 @@ def matrix_rank(A, tol=None, hermitian=False, *, rtol=None):
+
     if rtol is None:
         rtol = max(A.shape[-2:]) * finfo(S.dtype).eps
+    rtol = broadcast_to(rtol, A.shape[:-2])[..., newaxis]
```
This should be inside the `if tol is None` branch. Indeed, the `if rtol is None` statement should be moved there - no point calculating `rtol` if it is not going to be used anyway.

mtsokol replied:
I changed it to `rtol = asarray(rtol)[..., newaxis]` as we still want to call `asarray` (similarly `asarray(tol)` later) to accept list inputs, like: `rtol = [0.2, 0.3, 0.1]`.

mhvk commented on numpy/linalg/tests/test_linalg.py:
```diff
@@ -1620,6 +1620,7 @@ def test_matrix_rank(self):
         assert_equal(matrix_rank(ms), np.array([3, 4, 0]))
         # works on scalar
         assert_equal(matrix_rank(1), 1)
+        assert_equal(matrix_rank(I, tol=0.0), matrix_rank(I, rtol=0.0))
```
I'm fairly sure this would not have failed without the fix - would suggest using the explicit example from the issue.

mtsokol replied:
Sure! I added a regression test with an example from the issue.

mhvk reviewed (commented):
A few small comments...

mtsokol commented:
> This looks like it fixes the problem, although I'm not sure if the broadcast_to is necessary.
> 
> Should we add tests?

I changed `broadcast_to` to `asarray` to accept list inputs also (`asarray(tol)` is already present).
I added a regression test.

mhvk commented on numpy/linalg/_linalg.py:
```diff
@@ -2068,18 +2068,23 @@ def matrix_rank(A, tol=None, hermitian=False, *, rtol=None):
-    if rtol is not None and tol is not None:
-        raise ValueError("`tol` and `rtol` can't be both set.")
+
     if rtol is None:
```
Could you still move this and the next two lines inside the `if tol is None` branch? Also, since in this branch, `rtol` is guaranteed to be a scalar, it can be,
```
if tol is None:
    if rtol is None:
        rtol = max(A.shape[-2:]) * finfo(S.dtype).eps
    else:
        rtol = asarray(rtol)[..., newaxis]
    tol = S.max(axis=-1, keepdims=True) * rtol
else:
    ...
```

mtsokol replied:
Ah right, I confused `tol` with `rtol` in your previous comment - now I should be right.

mhvk reviewed (approved):
OK, thanks for the quick change. Looks all good now!
